File: Mikel-Inaki-Victor/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    private final Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {

        for (int i = 0; i < items.length; i++) {
            items[i] = ItemFactory.createItem(items[i]).updateItem();
        }
    }
}

File: jrogalan_csharp/src/GildedRose/GildedRose.cs
using System.Collections.Generic;
using GildedRose.Core.Extensions;
using GildedRose.Entities;
using GildedRose.Policies;
using Genesis.Ensure;

namespace GildedRose
{
    public class GildedRose
    {
        readonly IPolicyFactory _factory;
        readonly IList<Item> _items;
        
        public GildedRose(IList<Item> items)
        {
            Ensure.ArgumentNotNull(items, nameof(items));

            _items = items;
            _factory = new PolicyFactory();
        }

        public void UpdateQuality() => 
            _items.ForEach(item => 
                _factory.Create(item).UpdateQuality());
    }
}